rdrf/report/views.py

```diff
@@ -0,0 +1,93 @@
+from django.contrib import messages
+from django.core.exceptions import PermissionDenied
+from django.http import StreamingHttpResponse
+from django.shortcuts import render, get_object_or_404, redirect
+from django.views import View
+
+from rdrf import settings
+from rdrf.security.mixins import SuperuserRequiredMixin
+from report import __version__
+from report.forms import ReportDesignerForm
+from report.models import ReportDesign
+from report.reports.generator import Report
+
+
+class ReportsAccessCheckMixin:
+    def dispatch(self, request, *args, **kwargs):
+        if not ReportDesign.objects.reports_for_user(request.user).filter(pk=kwargs['report_id']).exists():
+            raise PermissionDenied
+        return super().dispatch(request, *args, **kwargs)
+
+
+class ReportsView(View):
+    def get(self, request):
```

Comment:
My view is that reports can target Patients, therefore they should be able to see the page. 

It might be confusing for the user if they expected to be able to access the reports list page (had there been a report available to them), but then they get Permission Denied if that specific report got taken away. 

Otherwise if you want me to change it I need to add another check to determine whether to show Permission Denied. 
i.e.
```
reports = get reports for user
if user has no reports AND is not a superuser
   raise PermissionDenied
```

Let me know if you'd like me to proceed with the above.

Reply:
In most of trrf, there is clear separation between admin / staff functionality and patient functionality (admin & explorer pages), that is enforced at the 'response status' level. Previously, this enforcement was handled differently in different parts of the application (often mixed into the response logic), which lead to unexpected security vulnerabilities because it wasn't easy for devs to evaluate whether a view had the correct permissions unless you inspected the application logic.

In this PR, it is currently not possible for patients to access any reports anyway, because `reports_for_user` still requires that the user is a superuser, curator, or clinician. The reports page is also not visible to patients in the quicklinks.

It's possible we could allow patients and carers to have access to reports on themselves, but that would require a good usecase and more testing.

The `PermissionDenied` in the mixin wouldn't check whether any reports have been made available to them, just that they're the type of user that is expected to have access to reports.

Reply:
Ok, in that case, I will remove those groups from the UI as well.

![image](https://user-images.githubusercontent.com/60163872/161471052-4d02ac5d-962f-4f13-8c9e-73c55aa5e73c.png)

Reply:
Done.

Reply:
I think that change allows users that aren't in the desired access group to download a report in `ReportDownloadView`.

Reply:
oops. I've split it into 2 different checks now.